scheduled-cloudfunction/main.tf

```diff
@@ -1,49 +1,55 @@
-resource "google_pubsub_topic" "scheduled_function_pubsub" {
-  name = var.scheduled_function_name
-}
-
-resource "google_cloudfunctions_function" "scheduled-cloudfunction" {
-  name                  = var.scheduled_function_name
-  description           = var.scheduled_function_description
-  runtime               = var.function_runtime
-  ingress_settings      = "ALLOW_INTERNAL_ONLY"
-  service_account_email = var.service_account_email
-  timeout               = var.function_timeout
-  min_instances         = var.min_instances
-  max_instances         = var.max_instances
-  labels = {
-    managed-by = "terraform"
-  }
-
-  available_memory_mb = var.memory_mb
-  entry_point         = var.function_entrypoint
-
-  event_trigger {
-    event_type = "google.pubsub.topic.publish"
-    resource   = google_pubsub_topic.scheduled_function_pubsub.id
-  }
-
-  environment_variables = var.function_environment_variables
-
-  source_repository {
-    url = var.source_repository_url
-  }
-
-  lifecycle {
-    ignore_changes = [
-      labels["deployment-tool"]
-    ]
-  }
+resource "google_pubsub_topic" "scheduled_function_trigger_topic" {
+  name    = var.scheduled_function_name
+  project = var.project_id
 }
 
 resource "google_cloud_scheduler_job" "cloud_scheduler_schedule" {
   name        = var.scheduled_function_name
+  project     = var.project_id
   description = "Cron schedule for: ${var.scheduled_function_name}"
   schedule    = var.cron_schedule
   time_zone   = "America/New_York"
 
   pubsub_target {
-    topic_name = google_pubsub_topic.scheduled_function_pubsub.id
+    topic_name = google_pubsub_topic.scheduled_function_trigger_topic.id
     data       = base64encode("pleaserun")
   }
 }
+
+resource "google_service_account" "scheduled_function_service_account" {
```

Comment:
Is there a use case where we wouldn't want to create a service account for a function and would instead want to use one that already exists?

Reply:
I've bounced back and forth on this a bit. The original thinking was that it was better to leave service account creation up to the user. On one hand, this is nice because it allows for potentially sharing service accounts for functions with identical permissions.

On the other hand, in practice this has resulted in a lot of service account management outside the module, since the functions often need different sets of perms.

I can see maybe a best of both worlds scenario here where we either create the service account, or allow for specifying one. This kind of logic is not what terraform is particularly good at 😄 -- I'll play around with it and see if I can come up with something that doesn't resemble fiendish devilry.

Reply:
OK, I still need to test it to see if this works, but I just pushed an update that should allow for conditionally managing the service accounts.

If you enable it, it manages service accounts and permissions for you. Disabled, all the IAM and workload identity bits should be skipped, and you're on your own.

I can see there being a hybrid here, but maybe with more complexity in the ternary operators:

1. have input variables for the runtime and deployment service account emails
2. if the length of the strings providing these inputs is > 0, use that as the target for any IAM permissions / workload identity federation mappings we assign
3. if the length of the strings providing these inputs is 0, create the service accounts and use those created ones as the target for IAM permissions and workload identity federation mappings

Do either of these two approaches sound more appealing?

Reply:
having thought about this a little bit more, i think it's just easier to ask for/require service account IDs as inputs, and then assign the necessary permissions in the module. This will give the freedom to manage the service accounts elsewhere, and is only like ~4 lines of terraform outside the module if you need to create a function specific service account.